OpenStack Content Delivery V1 support (SDK V2)

sharwell opened the pull request:
# Work in Progress

This pull request adds support for the basic functionality of the OpenStack Content Delivery Service V1 to the V2 SDK.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <inheritdoc/>
+        public virtual Task<GetHealthApiCall> PrepareGetHealthAsync(CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
:exclamation: This needs to be implemented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <inheritdoc/>
+        public virtual Task<GetSubsystemHealthApiCall> PrepareGetSubsystemHealthAsync(SubsystemId subsystemId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
:exclamation: This needs to be implemented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <inheritdoc/>
+        public virtual Task<UpdateServiceApiCall> PrepareUpdateServiceAsync(ServiceId serviceId, ServiceData serviceData, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
:exclamation: This needs to be implemented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <inheritdoc/>
+        public virtual Task<RemoveAssetApiCall> PrepareRemoveAssetAsync(ServiceId serviceId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
:exclamation: This needs to be implemented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <inheritdoc/>
+        public virtual Task<GetFlavorApiCall> PrepareGetFlavorAsync(FlavorId flavorId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
:exclamation: This needs to be implemented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <inheritdoc/>
+        public virtual Task<RemoveFlavorApiCall> PrepareRemoveFlavorAsync(FlavorId flavorId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
:exclamation: This needs to be implemented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryServiceExtensions.cs:
```diff
@@ -0,0 +1,158 @@
+                task => task.Result.SendAsync(cancellationToken).Select(innerTask => innerTask.Result.Item2));
+        }
+
+        //public static Task<GetHealthApiCall> GetHealthAsync(this IContentDeliveryService service, CancellationToken cancellationToken)
```
:exclamation: These need to be implemented, or removed.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryServiceExtensions.cs:
```diff
@@ -0,0 +1,158 @@
+    /// <preliminary/>
+    public static class ContentDeliveryServiceExtensions
+    {
+        public static Task<HomeDocument> GetHomeAsync(this IContentDeliveryService service, CancellationToken cancellationToken)
```
:exclamation: This needs to be documented (along with several other cases in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/FlavorData.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+        }
+
+        public ImmutableArray<FlavorProvider> Providers
```
:exclamation:  This needs to be documented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/IContentDeliveryService.cs:
```diff
@@ -0,0 +1,53 @@
+    {
+        #region Service Information
+
+        Task<GetHomeApiCall> PrepareGetHomeAsync(CancellationToken cancellationToken);
```
:exclamation: This needs to be documented (along with several other cases in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/Service.cs:
```diff
@@ -0,0 +1,77 @@
+    using OpenStack.Services.Identity;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class Service : ServiceData
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceCache.cs:
```diff
@@ -0,0 +1,73 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceCache : ExtensibleJsonObject
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceCacheRule.cs:
```diff
@@ -0,0 +1,51 @@
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    public class ServiceCacheRule : ExtensibleJsonObject
```
:exclamation: This needs a `[JsonObject]` attribute.
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceData.cs:
```diff
@@ -0,0 +1,113 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceData : ExtensibleJsonObject
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceDomain.cs:
```diff
@@ -0,0 +1,66 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceDomain : ExtensibleJsonObject
```
:exclamation: This needs to be documented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceError.cs:
```diff
@@ -0,0 +1,34 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceError : ExtensibleJsonObject
```
:exclamation: This needs to be documented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceError.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+        }
+
+        public string Message
```
:exclamation: This needs to be documented.

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceOrigin.cs:
```diff
@@ -0,0 +1,83 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceOrigin : ExtensibleJsonObject
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceOriginRule.cs:
```diff
@@ -0,0 +1,51 @@
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    public class ServiceOriginRule : ExtensibleJsonObject
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceRestriction.cs:
```diff
@@ -0,0 +1,53 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceRestriction : ExtensibleJsonObject
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceRestrictionRule.cs:
```diff
@@ -0,0 +1,52 @@
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceRestrictionRule : ExtensibleJsonObject
```
:exclamation: This needs to be documented (along with several other elements in the same file).

sharwell commented on src/OpenStackNetTests.Live/ContentDeliveryTests.Shared.cs:
```diff
@@ -0,0 +1,405 @@
+﻿namespace OpenStackNetTests.Live
```
:exclamation: This file is currently only included in the Live tests project. It should be linked to the Unit tests project so a service simulator can be implemented for efficient unit testing of the service functionality.

DonSchenck commented:
I'm going to work on implementing the above methods.

sharwell commented:
@DonSchenck Awesome. I look forward to seeing the update! :+1:

DonSchenck commented:
Why is this tagged as Do Not Merge? Let's merge it and then I can pull it down and continue work.

sharwell commented:
In the work so far on the V2 SDK, we've only merged pull requests once they reach a certain level of completeness. In particular, while the functionality could be a subset of the complete offerings of the service, we looked for the following:

1. Implementations of code in the service client class.
2. Reasonably complete documentation.

We also looked for a minimal level of live testing coverage (focused on the most important API features).

There are two ways for multiple users to collaborate on a feature in this environment. To date, we've been using the first method.

1. The user who starts the work creates a feature branch in their fork of this repository. In the case of this pull request, the branch was created as **sharwell/v2-content-delivery**. Other users working on the same feature would add a new remote for the fork of the repository where the feature branch is. For example, the following adds a remote, checks out the feature branch, and sets your copy of the branch to upload to *your* fork of the project from which you can send a pull request:

    ```bash
    # Add a remote 'sharwell', where the feature branch is located (only need to do this once
    # per remote)
    git remote add sharwell https://github.com/sharwell/openstack.net.git
    # Fetch from the 'sharwell' remote, so your local copy has up-to-date information about its
    # working branches
    git fetch sharwell
    # Check out a feature branch from the 'sharwell' remote
    git checkout -b v2-content-delivery sharwell/v2-content-delivery
    # Set your local copy of the feature branch to track 'origin' remote instead of 'sharwell'.
    # This will push to your fork, and then you can submit a pull request to the original feature
    # branch in the 'sharwell' remote. When that pull request is merged into the feature branch,
    # *this* pull request (#475) will automatically be updated to include that work.
    git branch -u origin/v2-content-delivery v2-content-delivery
    ```

    To submit work to this feature branch prior to *this* pull request (#475) getting merged, a user can submit a pull request to the feature branch (see [Changing the branch range and destination repository](https://help.github.com/articles/using-pull-requests/#changing-the-branch-range-and-destination-repository) for more information on sending a pull request to a specific feature branch in a specific fork).

2. The alternative is to create feature branches in the parent repository. The workflow for this is essentially the same as for the first step, except users do not need to use `git remote add` and the intermediate pull requests only need to specify a branch (because the default repository is the correct target). The primary downside to this approach is it tends to lead to many branches in the main repository, which can be challenging for new users to sort through.

DonSchenck commented:
Got this:
```
C:\Users\User\Documents\GitHub\openstack.net [v2-content-delivery +1 ~1 -1 !]> git branch -u origin/v2-content-delivery v2-content-delivery
error: the requested upstream branch 'origin/v2-content-delivery' does not exist
hint:
hint: If you are planning on basing your work on an upstream
hint: branch that already exists at the remote, you may need to
hint: run "git fetch" to retrieve it.
hint:
hint: If you are planning to push out a new local branch that
hint: will track its remote counterpart, you may want to use
hint: "git push -u" to set the upstream config as you push.
```

sharwell commented:
@DonSchenck Thanks, I wasn't sure if that would happen. I changed that step to be two different commands in the original example. The `push` command I added at the end creates the remote branch mentioned in that error at the same time it assigns the tracking branch.

sharwell commented:
Superseded by #500.
